### PR TITLE
CSPL-1752: Do not update status on cached CR

### PR DESCRIPTION
**Pull request by sgontla**
On a reconcile entry, there is a possibility that the CR is cached(may not be
the latest). In such cases, CR status update fails, as the resource version
already changed.

Fix is made such that:
1. Always fetch the latest CR, then make the status update to avoid any update failures
2. Make sure that the cached CR is reflecting the latest, before yielding from current reconcile
so that the next reconcile has the latet status info to work with

### CODE REVIEW

**Review by vivekr-splunk — APPROVED** [review]

File: pkg/splunk/enterprise/util.go [hunk 1]
```diff
@@ -1948,3 +1948,121 @@ func migrateAfwFromPhase2ToPhase3(ctx context.Context, client splcommon.Controll
...
+			// may not be an issue(i.e., expectation is always to be declarative), but the  application
+			// specific status may not be idempotent(example. tryint to install an app which was already installed).
+			// So, always make sure that the cache is reflecting the latest CR, before the next event
+			// waiting in the Q triggers the next reconcile
+			for chkCnt := 0; chkCnt < maxRetryCountForCRStatusUpdate; chkCnt++ {
+				crAfterUpdate, err := fetchCurrentCRWithStatusUpdate(ctx, client, latestCR)
+				if err == nil && updatedCRVersion == crAfterUpdate.GetResourceVersion() {
+					scopedLog.Info("Cahe is reflecting the latest CR", "updated CR version", updatedCRVersion)
+					// Latest CR is reflecting in the cache
+					break
```
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/splunk/enterprise/util.go:1988]: As discussed, some reconciles may not have any chances, so, we are always depending on the update calls return version to check

File: pkg/splunk/enterprise/util.go [hunk 2]
```diff
@@ -1948,3 +1948,121 @@ func migrateAfwFromPhase2ToPhase3(ctx context.Context, client splcommon.Controll
...
+	var tryCnt int
+	for tryCnt = 0; tryCnt < maxRetryCountForCRStatusUpdate; tryCnt++ {
+		latestCR, err := fetchCurrentCRWithStatusUpdate(ctx, client, origCR)
+		if err != nil {
+			if origCR.GetDeletionTimestamp() == nil {
+				scopedLog.Error(err, "Unable to Read the latest CR from the K8s")
+			}
+
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/splunk/enterprise/util.go:1966]: As discussed, there are some cases where the deletion may span multiple reconciles, so continuing with the existing behavior, that way, the CR may be assumed to be available, till all the finalizers are gone.

[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/util.go:1966]: LGTM

File: pkg/splunk/enterprise/finalizers_test.go [hunk 3]
```diff
@@ -126,6 +126,23 @@ func splunkDeletionTester(t *testing.T, cr splcommon.MetaObject, delete func(spl
 					{MetaName: "*v1.PersistentVolumeClaim-test-splunk-pvc-stack1-var"},
 				}
 			}
+
+			switch cr.GetObjectKind().GroupVersionKind().Kind {
+			case "Standalone":
+				mockCalls["Get"] = append(mockCalls["Get"], spltest.MockFuncCall{MetaName: "*v3.Standalone-test-stack1"})
+				mockCalls["Get"] = append(mockCalls["Get"], spltest.MockFuncCall{MetaName: "*v3.Standalone-test-stack1"})
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/finalizers_test.go:133]: Just for the sake brevity and uniformity, I think another approach(probably better) can be like this - 
```
funcCalls := []spltest.MockFuncCall{
    spltest.MockFuncCall{MetaName: "*v3.Standalone-test-stack1"},
    spltest.MockFuncCall{MetaName: "*v3.Standalone-test-stack1"},
}
mockCalls["Get"] = append(mockCalls["Get"], funcCalls...)
```
What do you think? :)

File: pkg/splunk/enterprise/util.go [hunk 4]
```diff
@@ -1948,3 +1948,121 @@ func migrateAfwFromPhase2ToPhase3(ctx context.Context, client splcommon.Controll
...
+
+		scopedLog.Info("Trying to update", "count", tryCnt)
+		curCRVesion := latestCR.GetResourceVersion()
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/util.go:1969]: nit:curCRVersion

File: pkg/splunk/enterprise/util.go [hunk 5]
```diff
@@ -1948,3 +1948,121 @@ func migrateAfwFromPhase2ToPhase3(ctx context.Context, client splcommon.Controll
...
+		if err == nil {
+			updatedCRVersion := latestCR.GetResourceVersion()
+			scopedLog.Info("Status update successful", "current CR version", curCRVesion, "updated CR version", updatedCRVersion)
+
+			// While the current reconcile is in progress, there may be new event(s) from the
+			// list of watchers satisfying the predicates. That triggeres a new reconcile right after
+			// exiting from the current reconcile, in which case, refers the cached version of the
+			// CR missing the updates we are doing here. From K8s resource point of view, this
+			// may not be an issue(i.e., expectation is always to be declarative), but the  application
+			// specific status may not be idempotent(example. tryint to install an app which was already installed).
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/util.go:1980]: super duper nit: trying to install...

**Review by kashok-splunk — APPROVED** [review]

**Review by gaurav-splunk — APPROVED** [review]

**Pull request merged**